src/atomWithSearchParams.ts

```
@@ -0,0 +1,151 @@
+   */
+  const resolveValue = (value: string | null | undefined): T => {
+    // If the value is null, undefined, or not a string, return the default value.
+    if (value === null || value === undefined || typeof value !== 'string') {
+      return defaultValue;
+    }
+
+    try {
+      // Determine the type of the default value and parse accordingly.
+      if (typeof defaultValue === 'number') {
+        if (value === '') {
+          console.warn(
+            `Empty string provided for key "${key}". Falling back to default value.`,
+          );
+          return defaultValue;
+        }
+
+        const parsed = Number(value);
+        if (!Number.isNaN(parsed)) {
+          return parsed as T;
+        }
+
+        console.warn(`Expected a number for key "${key}", got "${value}".`);
+        return defaultValue;
+      }
+
+      // If the default value is a boolean, check if the value is `true` or `false`.
+      if (typeof defaultValue === 'boolean') {
+        if (value === 'true') return true as T;
+        if (value === 'false') return false as T;
+
+        console.warn(`Expected a boolean for key "${key}", got "${value}".`);
+        return defaultValue;
+      }
+
+      if (typeof defaultValue === 'string') {
+        return value as T;
+      }
+
+      // Handle object types (assume JSON if defaultValue is an object)
+      if (typeof defaultValue === 'object') {
+        const parsed = JSON.parse(value);
+        if (parsed && typeof parsed === 'object') {
+          return parsed as T;
+        }
+        console.warn(`Expected an object for key "${key}", got "${value}".`);
+        return defaultValue;
+      }
+    } catch (err) {
+      console.error(`Failed to resolve value for key "${key}":`, err);
+      return defaultValue;
+    }
+
+    // Fallback to default value for unsupported types
+    console.warn(`Unsupported defaultValue type for key "${key}".`);
+    return defaultValue;
```

Let's throw an error. We shouldn't reach here.